ENH: Enable custom compression levels in `np.savez_compressed`

a-sajjad72 opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

This pull-request adds **flexible compression support** to `np.savez_compressed` while preserving backward compatibility for existing code.


1. **Implementation**
   - Introduce a single **`zipfile_kwargs`** parameter that is forwarded verbatim to `zipfile.ZipFile`.  
     - Anything supported by the std-lib (`compression`, `compresslevel`, etc.) can now be passed through.  
     - Defaults to `compression=ZIP_DEFLATED` when `savez_compressed` is used and the caller does not specify one.
   - Human-readable aliases (`"stored"`, `"deflated"`, `"bzip2"`, `"lzma"`) are mapped case-insensitively to the appropriate `zipfile` constants.
   - Added validation for `compresslevel` ranges per algorithm and for invalid types.
   - The old reserved names `compression=` / `compression_opts=` are **removed**, so existing code that stores an array called `"compression"` will continue to work.

2. **Tests**
   - Re-worked `TestSavezCompressed` suite:  
     - Uses `zipfile_kwargs` everywhere.  
     - Covers basic use, algorithm/level permutations, error paths, mixed dtypes, file-handle / pathlib targets, and large-array “slow” tests.  
     - Adds new edge-case checks for bad `compression` / `compresslevel` types and out-of-range levels.  
   - All tests pass locally on CPython 3.11+ (NumPy’s current support floor).


## API changes

```python
np.savez_compressed(
    file, *arrays,
    allow_pickle=True,
    zipfile_kwargs=None,   # NEW – dict forwarded to zipfile.ZipFile
    **named_arrays,
)
```

* If `zipfile_kwargs` is `None`, it falls back to `{"compression": ZIP_DEFLATED}`.
* Example:

```python
np.savez_compressed(
    "data.npz",
    a=my_array,
    zipfile_kwargs={"compression": "lzma", "compresslevel": 9},
)
```


## Backward compatibility

   - No previously valid call is broken:
     - Calls without `zipfile_kwargs` continue to work unchanged.
     - Array names such as `"compression"` or `"compression_opts"` are no longer shadowed by function parameters.
   - Internal default remains ZIP_DEFLATED level 6 (as in the original implementation).


## Checklist

- [x] Core implementation in `_npyio_impl.py` & `zipfile_factory`.
- [x] Comprehensive test coverage updated in `TestSavezCompressed`.
- [x] Full test suite passes (`pytest -q`) on Linux/macOS/Windows, CPython 3.11+.
- [x] Documentation & docstrings updated.

a-sajjad72 committed:
ENH: Add support for compression options in `_savez` and `zipfile_factory`

a-sajjad72 committed:
ENH: Add tests for compressed savez functionality with various compression methods and options

a-sajjad72 committed:
STYLE: fix import ordering and whitespace linting issues in `_npyio_impl.py` and `test_io.py`

a-sajjad72 committed:
ENH: Validate integer compression constants and options in savez_compressed

ngoldbaum commented:
I shared this feedback over Slack but realized it makes more sense here.

Not sure why there’s code in this PR to support old python versions like python 3.3, we support Python 3.11 or newer.

It’s not clear to me why this feature is needed. NumPy is relatively conservative about adding new API surface, why add this?

You might want to ping the mailing list to get people to take a look. One of the contribution guidelines is to ping the mailing list before opening a PR adding a new feature.

a-sajjad72 committed:
Merge branch 'numpy:main' into compression-feat

a-sajjad72 committed:
Merge branch 'numpy:main' into compression-feat

a-sajjad72 committed:
Merge branch 'main' of https://github.com/numpy/numpy into compression-feat

a-sajjad72 committed:
REF: Update `savez_compressed` to use `zipfile_kwargs` for compression options

- Refactored `savez_compressed` to accept a `zipfile_kwargs` dictionary for compression settings, replacing the previous `compression` and `compression_opts` parameters.
- Updated related tests to utilize the new `zipfile_kwargs` structure for specifying compression methods and levels.
- Improved validation for compression levels and methods within the new structure.

a-sajjad72 committed:
Merge branch 'main' of https://github.com/numpy/numpy into compression-feat

a-sajjad72 committed:
REF: Enhance compression handling in `_savez` function

- Improved default behavior for compression settings, ensuring DEFLATED is used when compressing unless specified otherwise.
- Added support for translating string-based compression methods to their corresponding integer constants.
- Enhanced validation for `compresslevel`, ensuring it is an integer or None, and updated error messages for clarity.
- Refactored internal logic to streamline compression method handling and validation.

a-sajjad72 committed:
STY: Improve linting error messages and formatting in `_savez` and related tests

a-sajjad72 committed:
REF: Update compression tests and remove legacy Python checks

- Simplified exception handling for unavailable compression methods in tests.
- Removed legacy tests for Python versions <3.3, as NumPy now targets >=3.11.
- Added new tests for invalid compression types and levels, ensuring robust validation.
- Introduced case-insensitive handling for compression aliases in tests.

a-sajjad72 committed:
Merge branch 'main' of https://github.com/numpy/numpy into compression-feat

a-sajjad72 commented:
> So one reason we were a bit hesitant in the past, is that it'll break an array named `compression=...` in currently working code.
> 
> My gut feelig might be that a `zipfile_kwargs=dict` may make sense (weirder name, unless we add new API, and we would be clear that we forward it directly. That would still mean we promize to keep using `ZipFile`, but I suspect that is OK). (Yes, that is a bit awkward API wise and we don't have such API elsewhere in NumPy, so not sure.)

@seberg Thanks for pointing that out!  
I hadn’t considered that `compression=` can already be an array name.  
Now I had switched to a single `zipfile_kwargs` dict that we pass straight to `ZipFile`; that should avoid the conflict while still exposing the new functionality. I had pushed the changes have a review on it.

a-sajjad72 committed:
REF: Add stubs for `zipfile_kwargs` parameter of savez_compressed function

a-sajjad72 committed:
Merge branch 'main' into compression-feat

jorenham reviewed (approved):
the stub changes look good

a-sajjad72 committed:
ENH: Improve `savez_compressed` function by enhancing `zipfile_kwargs` handling and updating documentation. Added support for passing compression options directly to `zipfile.ZipFile` without validation. Updated tests to reflect changes in compression handling and error management.

a-sajjad72 committed:
revised proposed changes

a-sajjad72 committed:
revised proposed change
- dropped the test case

a-sajjad72 commented:
@seberg Just a friendly ping on this! 

to sumarize the latest state of the PR based on our last discussion:
- updated the `.pyi` stubs.
* now `zipfile_kwargs`  should avoid mutating the user's dictionary.
* regarding the `zipf.close()` test failure we discussed; since it seems to be a `zipfile` sanity-check bug, I dropped that specific test case(s) from the test suite so it no longer fails.

let me know if this approach looks good to you? or if there are any final tweaks needed before merging!
